File: ecommerce_reasoning.py
"""
Logical Reasoning for E-commerce Stores using Forward and Backward Chaining
"""

# Forward Chaining Example: Discount Recommendation
def forward_chaining_discount(user):
    facts = []
    if user['purchases'] > 5:
        facts.append('loyal_customer')
    if user['cart_value'] > 100:
        facts.append('high_cart')
    # Rules
    if 'loyal_customer' in facts and 'high_cart' in facts:
        return 'Recommend 10% discount'
    elif user['is_new']:
        return 'Recommend 5% discount'
    else:
        return 'No discount'

# Backward Chaining Example: Free Shipping Eligibility
def backward_chaining_free_shipping(user):
    # Goal: eligible for free shipping
    if user['cart_value'] > 200:
        return True
    elif user['is_premium']:
        return True
    else:
        return False

if __name__ == "__main__":
    user1 = {'purchases': 8, 'cart_value': 120, 'is_new': False, 'is_premium': False}
    user2 = {'purchases': 1, 'cart_value': 250, 'is_new': True, 'is_premium': True}
    print("Forward Chaining Discount for user1:", forward_chaining_discount(user1))
    print("Backward Chaining Free Shipping for user1:", backward_chaining_free_shipping(user1))
    print("Forward Chaining Discount for user2:", forward_chaining_discount(user2))
    print("Backward Chaining Free Shipping for user2:", backward_chaining_free_shipping(user2))

File: ecommerce_reasoning.pl
% Logical Reasoning for E-commerce Stores using Forward and Backward Chaining

% Facts: user(purchases, cart_value, is_new, is_premium)
user(user1, 8, 120, false, false).
user(user2, 1, 250, true, true).

% Forward Chaining Rules: Discount Recommendation
recommend_discount(User, '10%') :-
    user(User, Purchases, CartValue, _, _),
    Purchases > 5, CartValue > 100.
recommend_discount(User, '5%') :-
    user(User, _, _, true, _).
recommend_discount(User, 'No discount') :-
    user(User, _, _, _, _),
    \+ recommend_discount(User, _).

% Backward Chaining Rules: Free Shipping Eligibility
eligible_free_shipping(User) :-
    user(User, _, CartValue, _, _), CartValue > 200.
eligible_free_shipping(User) :-
    user(User, _, _, _, true).

% Example queries:
% ?- recommend_discount(user1, Discount).
% ?- eligible_free_shipping(user2).
